File: clients/cpp/src/macos.cpp
/// macOS implementation: uses the built-in `security` CLI via popen().
///
/// Stored entries use the macOS Keychain's "generic password" class.
///   Service: "keysync/global" or "keysync/project/<name>"
///   Account: the secret key name (e.g. "DATABASE_URL")

#ifdef __APPLE__

#include "internal_helpers.hpp"
#include "keysync/errors.hpp"
#include "keysync/credential.hpp"

#include <cstdio>
#include <cstdlib>
#include <memory>
#include <string>
#include <vector>

namespace keysync {
namespace internal {
namespace {

/// Execute a shell command with popen() and return stdout as string.
/// Returns empty string on failure.
std::string execRead(const char* cmd) {
    std::string result;
    FILE* pipe = popen(cmd, "r");
    if (!pipe) {
        return result;
    }

    char buffer[256];
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        result += buffer;
    }

    int status = pclose(pipe);
    (void)status; // suppress unused warning

    return result;
}

/// Extract a named attribute value from a `security dump-keychain` record.
/// Records look like:
///     "svce"<NULL>
///     "acct"<NULL>
/// or
///     "svce"<blob>="keysync/global"
///     "acct"<blob>="DATABASE_URL"
std::string findAttrValue(const std::string& record, const std::string& attrName) {
    std::string search = "\"" + attrName + "\"";
    auto idx = record.find(search);
    if (idx == std::string::npos) {
        return "";
    }

    std::string after = record.substr(idx + search.size());

    auto eqIdx = after.find('=');
    if (eqIdx == std::string::npos) {
        return "";
    }

    std::string val = after.substr(eqIdx + 1);
    // Trim whitespace
    auto start = val.find_first_not_of(" \t\n\r");
    if (start == std::string::npos) {
        return "";
    }
    val = val.substr(start);

    if (val == "<NULL>") {
        return "";
    }

    if (!val.empty() && val[0] == '"') {
        auto end = val.find('"', 1);
        if (end != std::string::npos) {
            return val.substr(1, end - 1);
        }
    }

    // Trim trailing quotes
    while (!val.empty() && val.back() == '"') {
        val.pop_back();
    }
    return val;
}

} // anonymous namespace

std::string getSecretMacOS(const std::string& service, const std::string& account) {
    // Build command: security find-generic-password -s <service> -a <account> -w 2>&1
    std::string cmd = "security find-generic-password -s \"";
    cmd += service;
    cmd += "\" -a \"";
    cmd += account;
    cmd += "\" -w 2>&1";

    std::string output = execRead(cmd.c_str());

    if (output.empty()) {
        throw KeySyncError(ErrorCode::NotFound,
            "secret not found: " + service + "/" + account);
    }

    // Check if output contains "security:" error prefix
    if (output.find("security: ") == 0) {
        // Exit code 44 means item not found, anything else is a keychain error
        if (output.find("could not be found") != std::string::npos ||
            output.find("The specified item could not be found") != std::string::npos) {
            throw KeySyncError(ErrorCode::NotFound,
                "secret not found: " + service + "/" + account);
        }
        throw KeySyncError(ErrorCode::KeychainError, trimTrailing(output));
    }

    return trimTrailing(output);
}

std::vector<CredentialEntry> listSecretsMacOS(const std::string& scope_filter,
                                               const std::string& project_filter) {
    std::vector<CredentialEntry> entries;

    // Dump the keychain and parse generic password entries
    std::string output = execRead("security dump-keychain 2>&1");

    if (output.empty()) {
        return entries;
    }

    // Split on "keychain:" records
    size_t pos = 0;
    while (true) {
        auto nextRecord = output.find("\nkeychain:", pos);
        std::string record;

        if (nextRecord == std::string::npos) {
            record = output.substr(pos);
        } else {
            record = output.substr(pos, nextRecord - pos);
        }

        // Only process generic password ("genp") entries
        if (record.find("class: \"genp\"") != std::string::npos) {
            std::string svc = findAttrValue(record, "svce");
            if (svc.size() >= 8 && svc.substr(0, 8) == "keysync/") {
                std::string acct = findAttrValue(record, "acct");
                if (!acct.empty()) {
                    std::string entryScope, entryProject;
                    parseServiceName(svc, entryScope, entryProject);

                    bool scopeMatch = scope_filter.empty() || entryScope == scope_filter;
                    bool projectMatch = project_filter.empty() || entryProject == project_filter;

                    if (scopeMatch && projectMatch) {
                        entries.push_back({entryScope, entryProject, acct});
                    }
                }
            }
        }

        if (nextRecord == std::string::npos) {
            break;
        }
        pos = nextRecord + 1; // skip past \n
    }

    return entries;
}

bool isNotFoundMacOS(const std::string& service, const std::string& account) {
    std::string cmd = "security find-generic-password -s \"";
    cmd += service;
    cmd += "\" -a \"";
    cmd += account;
    cmd += "\" -w 2>&1";

    std::string output = execRead(cmd.c_str());

    if (output.empty()) {
        return true;
    }

    // security returns exit code 44 for not found
    if (output.find("could not be found") != std::string::npos ||
        output.find("The specified item could not be found") != std::string::npos) {
        return true;
    }

    return false;
}

} // namespace internal
} // namespace keysync

#endif // __APPLE__

File: clients/rust/src/macos.rs
use std::process::Command;

use crate::error::{KeySyncError, Result};
use crate::credential::CredentialEntry;
use crate::service::parse_service_name;

/// Retrieve a secret from the macOS Keychain via the `security` CLI.
pub(crate) fn get_secret(service: &str, account: &str) -> Result<String> {
    let output = Command::new("security")
        .args([
            "find-generic-password",
            "-s", service,
            "-a", account,
            "-w",
        ])
        .output()?;

    // macOS 13+ may write the password to stderr instead of stdout.
    // Check both streams and use whichever is non-empty.
    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();

    // On some macOS versions, `security` returns the error message as stdout
    // with exit code 0 when the item is not found, instead of exit code 44.
    // Detect this and return NotFound.
    let not_found_msg = "SecKeychainSearchCopyNext: The specified item could not be found in the keychain.";
    if stdout.contains(not_found_msg) || stderr.contains(not_found_msg) {
        return Err(KeySyncError::NotFound);
    }

    if !stdout.is_empty() {
        return Ok(stdout);
    }
    if !stderr.is_empty() {
        return Ok(stderr);
    }

    if !output.status.success() {
        // security returns exit code 44 when the item is not found
        if output.status.code() == Some(44) {
            return Err(KeySyncError::NotFound);
        }
        return Err(KeySyncError::KeychainError(
            format!("security find-generic-password exited with status {}",
                output.status)
        ));
    }

    Ok(String::new())
}

/// List all keysync secrets by parsing the keychain dump.
pub(crate) fn list_secrets() -> Result<Vec<CredentialEntry>> {
    let output = Command::new("security")
        .arg("dump-keychain")
        .output()?;

    let out = String::from_utf8_lossy(&output.stdout);
    let records: Vec<&str> = out.split("\nkeychain:").collect();

    let mut entries = Vec::new();
    for rec in records {
        let rec = rec.trim();
        if rec.is_empty() || !rec.contains("class: \"genp\"") {
            continue;
        }

        let svc = find_attr_value(rec, "svce");
        if svc.is_empty() || !svc.starts_with("keysync/") {
            continue;
        }

        let acct = find_attr_value(rec, "acct");
        if acct.is_empty() {
            continue;
        }

        let (scope, project, env) = parse_service_name(svc);
        entries.push(CredentialEntry::new(scope, project, env, acct));
    }

    Ok(entries)
}

/// Returns true if the error indicates the secret was not found.
pub(crate) fn is_not_found(err: &KeySyncError) -> bool {
    matches!(err, KeySyncError::NotFound)
}

/// Extract an attribute value from a dump-keychain record line.
fn find_attr_value<'a>(record: &'a str, attr_name: &str) -> &'a str {
    let needle = format!("\"{}\"", attr_name);
    let idx = match record.find(&needle) {
        Some(i) => i,
        None => return "",
    };

    let after = &record[idx + needle.len()..];
    let eq_idx = match after.find('=') {
        Some(i) => i,
        None => return "",
    };

    let val = after[eq_idx + 1..].trim();
    if val == "<NULL>" {
        return "";
    }

    if val.starts_with('"') {
        let inner = &val[1..];
        if let Some(end) = inner.find('"') {
            return &val[1..end + 1];
        }
    }

    val.trim_matches('"')
}
